### PR TITLE
feat: Enhancement in subscription

### PR DESCRIPTION
**Pull request by deepeshgarg007**
image

- Creation of subscription against Suppliers
- Create Invoices even if subscription is Unpaid or Past Due
- Create invoices based on calendar months

Docs: https://github.com/frappe/erpnext_documentation/pull/101

### CODE REVIEW

File: erpnext/accounts/doctype/subscription/subscription.json [hunk 1]
```diff
@@ -195,10 +186,61 @@
...
+   "fieldname": "purchase_tax_template",
+   "fieldtype": "Link",
+   "label": "Purchase Taxes and Charges Template",
+   "options": "Purchase Taxes and Charges Template"
+  },
+  {
+   "default": "0",
+   "fieldname": "follow_calendar_months",
+   "fieldtype": "Check",
+   "label": "Follow Calendar Months",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.json:221]: Write a description of the field

File: erpnext/accounts/doctype/subscription/subscription.json [hunk 2]
```diff
@@ -195,10 +186,61 @@
...
+   "set_only_once": 1
+  },
+  {
+   "fieldname": "end",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.json]: end_date

File: erpnext/accounts/doctype/subscription/subscription.json [hunk 3]
```diff
@@ -195,10 +186,61 @@
...
+   "fieldtype": "Date",
+   "label": "Subscription End Date",
+   "set_only_once": 1
+  },
+  {
+   "default": "0",
+   "description": "New invoices will be generated as per schedule even if current invoices are unpaid or past due date",
+   "fieldname": "generate_new_invoices_past_due_date",
+   "fieldtype": "Check",
+   "label": "Generate New Invoices Past Due Date"
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.json:229]: Generate New Invoices On Due Date

File: erpnext/accounts/doctype/subscription_plan/subscription_plan.json [hunk 4]
```diff
@@ -61,15 +62,15 @@
    "fieldname": "price_determination",
    "fieldtype": "Select",
    "label": "Price Determination",
-   "options": "\nFixed rate\nBased on price list",
+   "options": "\nFixed rate\nBased on price list\nMonthly rate",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription_plan/subscription_plan.json]: The options should be in Title case, patch to fix existing

File: erpnext/accounts/doctype/subscription_plan/subscription_plan.json [hunk 5]
```diff
@@ -138,7 +139,8 @@
    "fieldtype": "Column Break"
   }
  ],
- "modified": "2019-07-25 18:35:04.362556",
+ "links": [],
+ "modified": "2020-06-14 22:39:48.732956",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription_plan/subscription_plan.json]: Fix roles for subscription plan

File: erpnext/accounts/doctype/subscription_plan/subscription_plan.json [hunk 6]
```diff
@@ -61,15 +62,15 @@
    "fieldname": "price_determination",
    "fieldtype": "Select",
    "label": "Price Determination",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription_plan/subscription_plan.json]: Subscription Price Based On

File: erpnext/accounts/doctype/subscription_settings/subscription_settings.json [hunk 7]
```diff
@@ -1,179 +1,66 @@
...
-   "role": "Administrator", 
-   "set_user_permissions": 0, 
-   "share": 1, 
-   "submit": 0, 
+   "create": 1,
+   "delete": 1,
+   "email": 1,
+   "print": 1,
+   "read": 1,
+   "role": "Administrator",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription_settings/subscription_settings.json]: Remove administrator
Add accounts manager and accounts user role

File: erpnext/accounts/doctype/subscription_invoice/subscription_invoice.json [hunk 8]
```diff
@@ -1,73 +1,40 @@
...
- "track_seen": 0
+ ],
+ "istable": 1,
+ "links": [],
+ "modified": "2020-06-01 22:23:54.462718",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "Subscription Invoice",
+ "owner": "Administrator",
+ "permissions": [],
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription_invoice/subscription_invoice.json:35]: Fix permissions

File: erpnext/accounts/doctype/subscription/test_subscription.py [hunk 9]
```diff
@@ -517,10 +531,12 @@ def test_prepaid_subscriptions_with_prorate_true(self):
 		settings.save()
 
 		subscription = frappe.new_doc('Subscription')
-		subscription.customer = '_Test Customer'
+		subscription.party_type = 'Customer'
+		subscription.party = '_Test Customer'
 		subscription.generate_invoice_at_period_start = True
 		subscription.append('plans', {'plan': '_Test Plan Name', 'qty': 1})
 		subscription.save()
+		subscription.process()
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/test_subscription.py:556]: Add new test cases for "follow_calendar_months" and  "generate_new_invoices_past_due_date"

File: erpnext/accounts/doctype/subscription/subscription.py [hunk 10]
```diff
@@ -469,10 +592,7 @@ def get_all_subscriptions():
 	"""
 	Returns all `Subscription` documents
 	"""
-	return frappe.db.sql(
-		'select name from `tabSubscription` where status != "Cancelled"',
-		as_dict=1
-	)
+	return frappe.db.get_all('Subscription', {'status': 'Cancelled'})
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.py]: status not cancelled

File: erpnext/accounts/doctype/subscription/subscription.py [hunk 11]
```diff
@@ -337,17 +435,18 @@ def is_prepaid_to_invoice(self):
...
+		invoice = self.get_current_invoice()
 
-		return False
+		if invoice:
+			if getdate(self.current_invoice_start) <= getdate(invoice.posting_date) <= getdate(self.current_invoice_end):
+				return True
+			else:
+				False
+		else:
+			return False
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.py]:
```suggestion
		if invoice and getdate(self.current_invoice_start) <= getdate(invoice.posting_date) <= getdate(self.current_invoice_end):
			return True
		return False
```

File: erpnext/accounts/doctype/subscription/subscription.py [hunk 12]
```diff
@@ -180,22 +212,24 @@ def current_invoice_is_past_due(self, current_invoice=None):
 		if not current_invoice:
 			current_invoice = self.get_current_invoice()
 
-		if not current_invoice:
+		if not current_invoice or self.is_not_outstanding(current_invoice):
```
Comment (nabinhait) [inline erpnext/accounts/doctype/subscription/subscription.py]: `is_not_outstanding` can be renamed to `is_paid`

**Review by nabinhait — CHANGES_REQUESTED** [review]
update docs

File: erpnext/accounts/doctype/subscription/subscription.json (hunk 3 above)
Comment (deepeshgarg007) [reply erpnext/accounts/doctype/subscription/subscription.json:229]: New invoices will be generate past 'Due Date' and not on 'Due Date' Itself so I feel past due date makes more sense here, rest changes I have made

File: erpnext/accounts/doctype/subscription_plan/subscription_plan.json (hunk 6 above)
Comment (nabinhait) [reply erpnext/accounts/doctype/subscription_plan/subscription_plan.json]: ?

**Review by nabinhait — CHANGES_REQUESTED** [review]
- fix travis
- change label of Price Determination

**Comment by deepeshgarg007** [comment]
> * change label of Price Determination

image

**Review by nabinhait — APPROVED** [review]

**Pull request merged**